File: crypto_agent/tools/get_portfolio.py
import json
from .registry import register_tool


@register_tool(
    name="get_portfolio",
    description="Get complete portfolio overview: asset balances, open positions with PnL, and open orders.",
    schema={"type": "object", "properties": {}},
)
async def handle_get_portfolio(exchange, **_) -> str:
    try:
        lines = []

        balance = await exchange.fetch_balance()
        if balance:
            lines.append("Asset     | Free         | Total")
            lines.append("-" * 40)
            for asset, info in sorted(balance.items()):
                lines.append(f"{asset:<9} | {info['free']:>12.4f} | {info['total']:>12.4f}")
        else:
            lines.append("No assets.")

        lines.append("")

        positions = await exchange.fetch_positions()
        if positions:
            lines.append("Symbol           | Side   | Size         | Entry        | Mark         | PnL")
            lines.append("-" * 85)
            for sym, pos in positions.items():
                side = pos.get("side", "long").upper()
                size = pos.get("contracts", pos.get("amount", 0))
                entry = pos.get("avg_entry_price", 0)
                mark = pos.get("current_price", 0)
                pnl = pos.get("unrealized_pnl", 0)
                leverage = pos.get("leverage")
                lev_str = f" {leverage}x" if leverage else ""
                lines.append(
                    f"{sym:<16} | {side:<6} | {size:>12.4f} | {entry:>12.2f} | "
                    f"{mark:>12.2f} | {pnl:>+10.2f}{lev_str}"
                )
        else:
            lines.append("No open positions.")

        lines.append("")

        if hasattr(exchange, "_orders"):
            orders = exchange._orders[-20:]
            if orders:
                lines.append("Recent Orders:")
                lines.append(json.dumps(orders, indent=2))
        else:
            open_orders = await exchange.fetch_open_orders()
            if open_orders:
                lines.append("Open Orders:")
                lines.append(json.dumps(open_orders, indent=2))

        return "\n".join(lines)
    except Exception as e:
        return f"Error: {e}"

File: src/tools/get-portfolio.ts
import { registerTool } from "./registry.js";
import { resolveToolTradingContext } from "./trading-context.js";

function formatUsdt(value: number): string {
  return value.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

registerTool(
  "get_portfolio",
  "Get complete portfolio overview: asset balances, open positions with PnL, and open orders.",
  { type: "object", properties: {} },
  ["exchange", "broker", "memory", "sessionId"],
  async ({ exchange, broker, memory, sessionId }) => {
    try {
      const lines: string[] = [];
      const ctx = memory ? resolveToolTradingContext(memory, sessionId) : null;
      const allocation = ctx ? memory?.getBotAllocation?.(ctx.botId, ctx.tradingAccountId, "USDT") : null;
      const balance = broker && ctx ? await broker.fetchBalance(ctx.botId) : await exchange.fetchBalance();

      if (ctx && allocation) {
        lines.push(`Active Bot: ${ctx.botId} (${ctx.tradingAccountId})`);
        lines.push(
          `Allocation USDT: allocated ${formatUsdt(allocation.allocated)} | free ${formatUsdt(allocation.free)} | used ${formatUsdt(allocation.used)} | realized PnL ${formatUsdt(allocation.realizedPnl ?? 0)}`,
        );
        lines.push("");
      }

      if (Object.keys(balance).length) {
        lines.push("Asset     | Free         | Total");
        lines.push("-".repeat(40));
        for (const [asset, info] of Object.entries(balance).sort()) {
          const i = info as any;
          lines.push(`${asset.padEnd(9)} | ${i.free.toFixed(4).padStart(12)} | ${i.total.toFixed(4).padStart(12)}`);
        }
      } else {
        lines.push("No assets.");
      }
      lines.push("");

      const positions = broker && ctx ? await broker.fetchPositions(ctx.botId) : await exchange.fetchPositions();
      if (Object.keys(positions).length) {
        lines.push("Symbol           | Side   | Size         | Entry        | Mark         | PnL");
        lines.push("-".repeat(85));
        for (const [sym, pos] of Object.entries(positions) as any) {
          const side = (pos.side ?? "long").toUpperCase();
          const size = pos.contracts ?? pos.amount ?? 0;
          const entry = pos.avg_entry_price ?? 0;
          const mark = pos.current_price ?? 0;
          const pnl = pos.unrealized_pnl ?? 0;
          const lev = pos.leverage ? ` ${pos.leverage}x` : "";
          lines.push(
            `${sym.padEnd(16)} | ${side.padEnd(6)} | ${size.toFixed(4).padStart(12)} | ${entry.toFixed(2).padStart(12)} | ${mark.toFixed(2).padStart(12)} | ${(pnl >= 0 ? "+" : "") + pnl.toFixed(2).padStart(9)}${lev}`,
          );
        }
      } else {
        lines.push("No open positions.");
      }
      lines.push("");

      if (exchange._orders) {
        const orders = exchange._orders.slice(-20);
        if (orders.length) {
          lines.push("Recent Orders:");
          lines.push(JSON.stringify(orders, null, 2));
        }
      } else {
        const openOrders = broker && ctx ? await broker.fetchOpenOrders(null, ctx.botId) : await exchange.fetchOpenOrders();
        if (openOrders.length) {
          lines.push("Open Orders:");
          lines.push(JSON.stringify(openOrders, null, 2));
        }
      }

      return lines.join("\n");
    } catch (e: any) {
      return `Error: ${e.message ?? e}`;
    }
  },
);
